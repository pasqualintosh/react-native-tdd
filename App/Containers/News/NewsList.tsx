import {Box, Button, FlatList, HStack, Text, VStack} from 'native-base';
import React from 'react';
import {getNewsItems} from '../../Utils/Api';
import NewsItem from './NewsItem';

export const getButtonTitle = (storiesType: 'TOP' | 'NEW') => {
  if (storiesType === 'TOP') return 'Show new stories';
  return 'Show top stories';
};

const NewsList: React.FC = (): JSX.Element => {
  const [newsItems, setNewsItems] = React.useState<Array<string>>([]);
  const [storiesType, setStoriesType] = React.useState<'TOP' | 'NEW'>('TOP');

  const toggleStoriesType = () => {
    setStoriesType(currentStoriesType =>
      currentStoriesType === 'TOP' ? 'NEW' : 'TOP',
    );
  };

  const intialize = async (): Promise<void> => {
    const ids = await getNewsItems();
    setNewsItems(ids.slice(0, 20));
  };

  React.useEffect(() => {
    intialize();
  }, []);

  return (
    <>
      <Box safeArea>
        <HStack justifyContent={'center'} alignItems={'center'}>
          <Button
            onPress={toggleStoriesType}
            size={'full'}
            testID="toggleStoriesButton">
            <Text>{getButtonTitle(storiesType)}</Text>
          </Button>
        </HStack>
        <VStack justifyContent={'center'} alignItems={'center'}>
          {newsItems.length > 0 && (
            <FlatList
              data={newsItems}
              renderItem={({item}) => <NewsItem id={item} />}
              keyExtractor={id => id}
              testID="listStories"
            />
          )}
        </VStack>
      </Box>
    </>
  );
};

export default NewsList;
