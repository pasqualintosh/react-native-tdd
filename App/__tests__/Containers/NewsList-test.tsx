import React from 'react';
import {cleanup, render, act, waitFor} from '@testing-library/react-native';

import NewsList from '../../Containers/News/NewsList';
import {NativeBaseProvider} from 'native-base';

jest.mock('./../../Utils/Api', () => {
  // const originalModule = jest.requireActual('./../../Utils/Api');
  return {
    __esModule: true,
    // ...originalModule,
    // default: jest.fn(() => ''),
    getNewsItems: jest.fn(() => {
      return [29874061, 29873106];
    }),
    getNewsItem: jest.fn(() => {
      return {id: 29874061, title: 'foo'};
    }),
  };
});

describe('NewList', () => {
  beforeEach(() => {});
  afterEach(() => cleanup());
  const inset = {
    frame: {x: 0, y: 0, width: 0, height: 0},
    insets: {top: 0, left: 0, right: 0, bottom: 0},
  };

  it('renders correctly', async () => {
    await act(async () => {
      const page = await waitFor(() =>
        render(
          <NativeBaseProvider initialWindowMetrics={inset}>
            <NewsList />
          </NativeBaseProvider>,
        ),
      );
      expect(page).toMatchSnapshot();
    });
  });

  it('loads news upon mount', async () => {
    await act(async () => {
      const root = await waitFor(() =>
        render(
          <NativeBaseProvider initialWindowMetrics={inset}>
            <NewsList />
          </NativeBaseProvider>,
        ),
      );

      return new Promise((resolve, reject) => {
        setTimeout(() => {
          expect(root.queryAllByTestId('listStories')).not.toMatchObject([]);
          resolve();
        }, 3000);
      });
    });
  });
});
