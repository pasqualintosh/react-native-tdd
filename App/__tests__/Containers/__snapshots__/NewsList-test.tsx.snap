// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewList renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={Object {}}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <View
        _disabled={
          Object {
            "bg": "trueGray.300",
            "opacity": "50",
          }
        }
        _loading={
          Object {
            "bg": "warmGray.50",
            "opacity": "50",
          }
        }
        _text={
          Object {
            "color": "lightText",
          }
        }
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        colorScheme="primary"
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onHoverIn={[Function]}
        onHoverOut={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#06b6d4",
            "borderRadius": 4,
            "flexDirection": "row",
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
        testID="toggleStoriesButton"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <View
            style={Object {}}
          >
            <Text
              style={
                Object {
                  "backgroundColor": undefined,
                  "color": "#1f2937",
                  "fontFamily": undefined,
                  "fontSize": 14,
                  "fontStyle": "normal",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                  "lineHeight": 21,
                  "textDecorationLine": undefined,
                }
              }
            >
              Show new stories
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    />
  </View>
</RNCSafeAreaProvider>
`;
